### DISCUSSION TITLE
PCB Design for potential additional stepper

### DISCUSSION THREAD

**Discussion by zyeborm** [kachurovskiy/nanoels#106] [Ideas]
Hi There.
My father (Russell) was wanting an ELS for his lathe, came across your project and liked it. He is more mechanically inclined and I am more electrical/software. He sent me your video where you had what looks like the nanoels running multiple axies during a threading operation. https://www.youtube.com/watch?v=B8pwe0wKsCU

I understand the software isn't there yet but PCB design is part of my day job (Not that I am an expert) and I was going to look at revisiting your current PCB to improve noise handling, ESD safety, things like that. I figured it may be nice at the same time to add provision for the extra stepper (and perhaps some general inputs/outputs, limit switches etc perhaps?). If it's ok could you tell me what pins you would like to use for the stepper (if they are special at all, eg if you are using hardware PWM for outputs or something) 
And any other things you'd like to see on your dream PCB ;-)

Just to be clear, not suggesting there are any faults or failures in the existing design. I want to test some things I have learnt about PCB design and this is a good project for that. 

I was probably going to look at getting boards assembled by JLC and using SMT parts, it's almost cheaper getting 5 boards made and assembled that way than it is buying components for a single board these days lol. That and I need some more practise ordering PCBA services.

Happy to give PCB designs back to you and your very nice project of course.

P.S.
I *may* be able to help a little with software as well, but that takes a lot more time as I'm sure you're aware. I haven't had a close look at your code but what I have seen looks nice. I think I may have spotted a few places where you could save some cpu cycles though if that's of any use to you? (Readability and "it's not broken" have a lot going for them though lol)

**Comment by kachurovskiy** [comment]
Hi @zyeborm, huge thanks for offering your help. Arduino is not the best platform for ELS and for 2-axis controller I've moved to ESP32 in the upcoming version of NanoEls. It has 15x higher frequency, 2 cores and many more features for a smaller price :) And it's supported by the Arduino IDE just as well.

I have the following PCB in production at the moment with ALLPCB (JLCPCB had component shortage):

![image](https://user-images.githubusercontent.com/517919/214777341-77650299-91dc-4398-86ba-a27374a77e35.png)

![image](https://user-images.githubusercontent.com/517919/214777456-296d2938-0aae-4abf-8c8e-8037f9f67c28.png)

It's designed to be used with a 25-button panel which I'm still trying to figure out that piece on because my last prototype has problems with buttons having 3 states instead of 2 (released, clicked & no contact, clicked & has contact). Doesn't help that my suppliers are all out of office due to Chinese New Year :)

![image](https://user-images.githubusercontent.com/517919/214777885-28ba7b63-1344-4230-b6ac-c4e7661a2f0b.png)

If you notice any problems with the designs above in terms of ESD or EMI, please let me know. The code for the 2-axis version will be open-source but I'm still thinking whether I'd be sharing the PCB design files since the plan is to sell those boards in my little online shop to recoup some of the investment for this project. Smallest batch of these (5 boards) costs 200$+ to produce and it'd be much cheaper for folks to buy a board from me than order it from JLCPCB. Same for the 25-button panel, minimal order there is 400$+ for 5 panels while I would be able to offer them for much less due to volumes.

I like your limit switch idea but on the ESP32 board we don't have any more free pins. Perhaps you could wire a NC limit switch into the driver/motor circuitry, not sure if it would be practical.

I don't plan to make more hardware iterations on H1/H2 Arduino boards but feel free to make your own hardware version of H2!

Let me know what you think, cheers!

**Reply by zyeborm** [reply]
Heh I'm up to my armpits in an esp32 project right now lol.
The new board design looks much better than the last one.
This is just commenting based on the PCB, it's hard to tell many things without the schematic.

It looks like stitched ground planes top and bottom? Unless you are a 4/6 layer board I'd probably go power top gnd bottom but it's much of a much ness. 
Honestly I'd probably go a 4/6 layer board these days given the cost difference and the EM environment you're in. I'd probably go Gnd Sig+pwr, sig+pwr, gnd for the stackup and bury your signal lines inside the the ground. 

I haven't worked with bare ESP32s only used the modules. For $4 or whatever for the 32-S3 or whatever the module makes nice for EMI compliance etc and you can use the wifi if you want (why you'd want to I'm not sure lol). It also gets you a nice board layout that is known to work well for the high speed and puts the CPU in a metal cage.
I'd suggest dropping the size of some of the bypass caps on the ESP itself and moving them right up as close as you can to the pins.
Drop a via next to the leg of the cap that is getting its power from the top plane, next to the via coming from the bottom plane. It makes the current loop shorter for really fast transients. Basically the vias should be in pairs for high speed.

Are those diodes TVS diodes?  (D13 etc?) or a switch matrix? Guessing switch matrix?

I would look to add some protection to your I/O lines. TVS diodes are a good start. Then if you can include a resistor to limit max current for outputs it'll make the board a lot more robust. Like even a 10R resistor drops the peak short circuit current from infinite to 300mA and 15 gives you 50mA which is probably a lot more survivable (Ideally set it to whatever absolute max is in the data sheet, but you have to watch the drop for anything that's actually pulling current, Like say an opto isolator. Gut feeling 15R would probably still be fine with those and and errant screwdriver or stray wire strand won't kill the board.

For any off board inputs put higher value resistors in. The input impedance of the ADC is like 50K I think? and in digital mode even higher so even a 5K resistor in series with the input line isn't going to bother anything. You can then add a capacitor after that to do some analogue filtering before it hits your chip. Like if you low pass filtered anything except the encoder at 1Mhz, and the encoder at even 3 Mhz (you can twiddle the value by changing the caps by hand if you want after the board is made for prototypes) then you will reject all wide band noise before it even gets to your chip.

For the usb d+ and d- lines I'd take them both out the top side of that plug so the length matches and run them as 50 ohm (I think for usb?) differential traces, it's not really going to matter but it's good practise. Watch out for powering your board from USB and from a power supply at the same time. They can fight and blow up your PC or the board. A diode from usb power to your board power can let you still power the board itself from USB (at 4.4 volts) which is very nice for development, and means if you have a machine connected and it puts out high voltage/whatever it won't fry your PC. It can wind up trying to use your USB port to power everything on the 5V bus but protecting from that is often a pain in the butt. Hmm actually looking at how you have those level shifted outputs it's probably not as hard to do as it sounds? Have the input to your 3.3v reg run through diodes from 5V in and from usb. Then have encoder power etc connected direct to 5V in. So the stepper outputs etc will only run when you have supplied 5V power but the rest of the board will run from USB and you can have both connected at the same time without issue.

I'd probably put the pullup resistors on the encoder pins right at the pin and then put a decoupling cap right next to it. Transients either from the encoder stepping or from noise will use those traces as antennas (and vice versa)

I'd very much put a decoupling cap and some bulk capacitance at the 5v output there and on the motor outputs. Not to provide clean supply for the encoder, but to kill off noise coming back from the encoder into your power supplies. Probably a ferrite bead  in line too. (They are SMT and a few cents)

You can possibly ferrite the step/dir/en pins as well to reduce noise coming back up those but you'd want to check frequencies and such.

Anyway, I really should get back to work, sorry for the wall of text + stream of consciousness there lol.

**Reply by kachurovskiy** [reply]
> you can use the wifi if you want

I specifically opted for the QFN48 to not have an antenna so that the device is not considered an intentional radiator. I have no need for WiFi / Bluetooth in the foreseeable future and I was hoping it could simplify some certifications one day :) Maybe it wasn't necessary.

> Are those diodes TVS diodes? (D13 etc?)

Yes, those are Bi-Directional TVS diodes for keypad ESD protection.

Thank you for a lot of good advice, I will consider it for the next version of the board 👍

**Reply by sikanderc21** [reply]
Dear sir can you provide me these controls i will be very thankful to you my whatsapp +966592391076

**Reply by kachurovskiy** [reply]
@sikanderc21 dear friend, sorry but I can't sell you anything at this time. If you have trouble building it yourself, you should be able to find someone in your city who can assemble the project for you for reasonable money. All the instructions are available in this repository. Hope it helps!

**Reply by sikanderc21** [reply]
Dear i have not problem to assemble this one i only problem for components can you let me how to buy all components pcb i will order from jlc pcb but when i goto order for parts with pcb the components not available can you prefer me any website where its components availability

**Reply by kachurovskiy** [reply]
@sikanderc21 on https://jlcpcb.com/ you order just the empty PCB board, without any components. It should cost around 20$ with delivery for 5 boards (minimum order).

On https://www.aliexpress.com/ you can order all the components:

- Arduino Nano (mini USB style)
- 2004 LCD display
- 12 buttons 6x6mm, each 6mm tall
- 8 1kOhm resistors, one 1MOhm resistor (or similar)
- 6 M3 nylon stand-offs (12mm long) and screws (come as a set in a box)
- 3 KF2EDG sockets (Curved needle, 4P, 3.5MM)
- 3 KF2EDG plugs (Plug Terminal, 4P, 3.5MM)
- MTS-102 on-off switch
- RV09 10kOhm potentiometer
- Set of M3 threaded inserts

Cheers!

**Comment by kachurovskiy** [comment]
I stand corrected, looks like I'm doing another iteration on the PCB design :) It's mostly to make it more robust and likely to pass the FCC/CE certifications (a man can dream).

<img width="1214" alt="r1" src="https://user-images.githubusercontent.com/517919/221807563-03e0a614-3ab3-46a2-9b67-ccbf7ec20894.PNG">

<img width="753" alt="r2" src="https://user-images.githubusercontent.com/517919/221807562-f7b00269-ceb0-45aa-bbda-e31c1e49c8d8.PNG">

<img width="751" alt="r3" src="https://user-images.githubusercontent.com/517919/221807553-e979f1ab-38c8-4885-a358-0ed7fa8c368a.PNG">

I tried to follow your recommendations where I understood them. Added a 300mA fuse even though the board only takes 100mA during normal operation. If there are any problems with this design, please kindly let me know :) Thanks!

**Comment by zyeborm** [comment]
Looks pretty good man ;-)

Somewhat unrelated I'm putting together a nanoels2 for my father right 
now (literally sat down to find the cable to program the arduino lol)
One thing I have found is the 3.5mm spacing headers you have for the 
stepper power etc is a little harder to find than 2.54 and 
5.8(something) spacing (I think, the physical store in Australia is 
Jaycar if you wanted to look lol).
Need to order from the big electronics supply place here, not just pick 
up from the local hobby store kind of thing. Not a deal breaker just 
something you might consider looking into?

Have you tried that low pass filter on the analogue button switch thing? 
That should breadboard ok to try out. I only suggest it because of the 
people who had noise issues.

You might consider getting your copper pour under the K2-K5 switches. 
But that's pretty much just a visual thing.


Oh, if you're getting it assembled consider liberally dousing the whole 
thing in LED's ;-)
They cost like less than a cent and a 10K(or so?, perhaps 5k? depending 
on LED) resistor driving is still plenty bright enough to see and now 
you can get an indication from the encoder that things are happening and 
also that you are step and directioning out.
I like the blinken lights (if they are dim lol) I'd breadboard that 
first to make sure it doesn't bother the encoder signals etc.
My father just got his encoder running and the Z (index) needed a pullup 
resistor to make a nice wave form? That seems very strange and LED's may 
impact on that.
Again, just an idea to consider for more flashing lights lol.


If you wanted to you could perhaps consider a switch matrix rather than 
using the analogue switches btw. It's a little more brain work to 
understand but it does give you a lot of digital switches with few I/O pins.


For your EMC dreams you might consider ferrite p.i networks at the feed 
into the nano. the EMI filter is a great start but take a look at the 
example circuit on page 2 of the datasheet and you'll see they have a 
cap on either side of it as well.
The resistors on your I/O lines will help with a lot in regards EMI too 
(slew rate limiting the signal edges, it drops the frequencies you put 
out dramatically.) I'd consider adding a ground contact (if you can) to 
the step/dir line outputs. People can run shielded lines then if they 
want (and you say they will for your testing lol)
I really wish the stepper drivers these days actually took differential 
inputs, it'd be so much nicer.
I'd perhaps look at "common mode" filtering and some capacitance right 
at the point that you get (and give) your 5V btw. You want to think 
about the wires leading into your board as being antennas, any ripple in 
the signal you put out over those antennas becomes radio waves and those 
wires are included when your device is tested. You can get a bit of an 
idea of your noise sources with a loop of wire on your oscilloscope (or 
spectrum analyser) and sniffing around your board with that. Won't tell 
you what passes/fails but you can get a magnitude comparison for any 
changes you make.

Probably test those 1K resistors on the outputs of the steppers, just 
make sure that isn't too much to make the drivers marginal.

Under your power switch, if you route that track from the arduino around 
the other side of the switch you can stop it from hugging those through 
holes, make for fewer ooopses when soldering.
On the LCD side I'd drop that line that runs next to the pads down some 
just to give yourself a bit more clearance there.
I think that's probably a general thing, in your DRC check, perhaps 
think about increasing the minimum clearance distance, like you're 
probably fine as is, but you have so much room on the board you might as 
well make it as easy to manufacture as possible? You can ignore it where 
you have to, but it might find places you're close but don't need to be.

Just to be really clear here I'm trying to find every possible thing I 
can see that might be wrong or could be improved (or made more to my 
personal taste lol) It's a good looking board and I don't think you'll 
have any particular problems with it.

Now I've done this instead of flashing my fathers PCB or building my 
mother in laws TV computer and now I'm late for bed too.
LOL, your project is way more fun that those things ;-)
Have a good day mate. Once I get this one up and running I may be able 
to help out a little more on the software side perhaps, see if we can 
steal some clock cylces back ;-)

**Reply by kachurovskiy** [reply]
Thanks a lot! That was helpful. Good luck with your build.

**Reply by zyeborm** [reply]
Just got the original H2 flashed, looks pretty good.
With 150K instead of 1M pull the voltages were off by like 3 counts and then only for a few of the switches lol.

Only new suggestion, take a look at the PCB footprint for that RV9 pot, it doesn't quite line up with the datasheet and it seems pretty loose when putting it in. Might be good for automated assembly and wave solder perhaps but doesn't seem right for hand assembly. 

Had a quick look through the code.
You might consider using the F macro in the nano serial prints. 
https://www.baldengineer.com/arduino-f-macro.html

Might save you some ram (if you were to wind up running low on it later, not needed now obviously lol)

**Reply by kachurovskiy** [reply]
> I only suggest it because of the people who had noise issues.

That user didn't use the PCB I think, just the breadboard.

> look at the PCB footprint for that RV9 pot, it doesn't quite line up with the datasheet

Sorry about that, I'm using a footprint someone contributed to JLCPCB library.

> You might consider using the F macro in the nano serial prints.

Nice one, didn't know about it. It shouldn't print anything after start unless you turn on the `DEBUG` mode though.

**Reply by zyeborm** [reply]
> > look at the PCB footprint for that RV9 pot, it doesn't quite line up with the datasheet
> 
> Sorry about that, I'm using a footprint someone contributed to JLCPCB library.

All good, was just letting you know about it as it looks like you are using the same one in this design.


> > You might consider using the F macro in the nano serial prints.
> 
> Nice one, didn't know about it. It shouldn't print anything after start unless you turn on the `DEBUG` mode though.

The way you have your code done with those #defines the debug stuff won't even be compiled so won't go into ram at all which is great.
Was just something I noticed you didn't have that can make a difference on the small micros.
I'm not sure if there's any speed hit to using it vs having them in ram, I know you do have tight timing requirements here.
Oh, I changed the serial rate to 115200 in mine btw. Especially as serial prints etc are blocking I try to get them out of the way ASAP lol.
It's not an uncommon speed in arduino projects I don't think and it's *so* much faster when you're printing lots of stuff during debugging etc.
On an ESP32 project I'm doing I have the serial rate set to 1mbit lol, just running over a 10cm ribbon cable to a USB>serial and it's working fine.

**Reply by kachurovskiy** [reply]
> Probably test those 1K resistors on the outputs of the steppers, just make sure that isn't too much to make the drivers marginal.

Well, it was too much :) Falling back to a ferrite bead and a ESD diode.

**Comment by kachurovskiy** [comment]
I'm thinking about adding some extra reverse polarity protection (Nano and screen already have some I believe) - do you think it's worth it? Will something like [this SBD](https://datasheet.lcsc.com/lcsc/1809140216_Jiangsu-Changjing-Electronics-Technology-Co---Ltd--B5819W-SL_C8598.pdf) work spec-wise?

Also thinking about over-voltage protection. Would a TVS diode to GND after the fuse be a good solution or is there a better way? Many thanks!

**Comment by zyeborm** [comment]
I don't think the nano has reverse polarity protection. Adding some is for sure a good idea especially to save peoples encoders etc, I think you should have enough voltage headroom to fit it in.

TVS diodes are more of a voltage spike than a sustained over voltage situation protection. Think ESD, sparks from walking across the carpet in socks kind of deal. You will want it to conduct until the fuse trips, they may well work but that could be a few seconds of operation if the input supply is like 6.something volts. (somebody turned the pot on the 5v supply)

Don't get me wrong it's probably a really good idea to put TVS on all the I/Os and the power just to help protect against ESD and big big noise. They say to put the TVS right on the edge but personally I wonder if it'd be better after your initial current limiting resistor, provided they are both located close to the actual pin. They want the TVS first to stop the spike from coupling into other things as it travels along traces. For this circuit you could probably get some USB TVS chips (they are pretty cheap and very common) to cover the I/O's that are directly connected to the nano. You get 2 channels out of each chip. Just check their clamping voltage, it might be too low being usb. 

For the overvoltage you could put a zener diode in after the fuse perhaps (and perhaps use a resettable/thermal fuse?). You'd want to check the curves but something like a 5.6 volt zener on the 5V rail probably won't eat too much current at idle and will probably trip your fuse at 6 ish volts? (ideally stay under the abs max voltage of the nano/screen)

If you really wanted to be fancy you can do the reverse polarity and over voltage stuff by putting a fet (or two, can't remember the circuit exactly) on your input rail and basically turn the fet off if the incoming voltage is too high or too low or backwards. 
There's a design in here https://electronics.stackexchange.com/questions/121670/zener-mosfet-overvoltage-protection that looks decent if you felt like going hard at it but it's probably overkill for the actual use of the device. (mostly overkill in time to get it designed, tested debugged etc)

But yeah a diode on the supply and possibly a zener (as a crowbar kinda) after the fuse is probably going to do the job good enough most of the time. I'd kinda like the crowbar(ish) zener to be after the diode  so you can use a lower voltage zener and keep closer to the limit of the arduino, If one exists like a 5.1v zener might do then. But then the reverse polarity diode has to take the short circuit current until the fuse trips and the zener itself might wind up chewing significant power all the time. Decisions Decisions lol. Thankfully this is your problem not mine lol.

**Reply by kachurovskiy** [reply]
Thank you! That makes sense. By the way, I've shared the 2-axis design in https://github.com/kachurovskiy/espina in case you're interested.

**Comment by zyeborm** [comment]
I haven't had a look at that one yet but I have had a bit more of a read of your code.
It looks like you're using the 16 but timer for step generation when you're in Async mode, but move to software generation when synchronised? What's the reason for that?
Especially given the impact it has on the rest of your system it seems like it'd be nicer to keep the step generation in hardware?

Some ideas for that, in your current Async ISR you're doing the step pulse in software still to get the delay. It's been a long time since I did avr timers but I think you should be able to allow the pwm to output directly.
I think if you set the pwm mode to fast pwm, and give it a set duty cycle (you might want inverted output) you can have the pwm output a pulse of a fixed duration.
The fun part is you can vary the frequency of those pulses by stuffing a start value into the timer register when the timer wraps add triggers the isr.
So in Async mode your step generation overhead becomes like just a few clock cycles to fill the register every step. 

If you also then in that isr kept track of the current position (++ or -- your position variable) you could change the synchronous operation from being dependant on the encoder pulses directly to a PID loop style of operation.

It wouldn't actually be that much of a change to your code I don't think because you wrote it so well. But basically use another timer or millis counter in your main loop to get a say 1khz trigger. (100hz would probably be totally fine too BTW)
In that task you see what your current position is, your desired position, then using the difference between those (the following error) update the value that the timer gets prefilled with to change the speed of the stepper to reduce the error.

Now you've got plenty of time to update LCDs and things, and your max step rate should be in the hundreds of khz range. (Back of the envelope says minimum step rate is about200hz if you're not using a prescaler, but what'll happen if you need to go below that is the pid loop will wind up basically dithering it's output provided the PID loop runs at more than the minimum step rate, it'll just put out a single step then run a few more loops and put it another.)

Double check everything I've said there though. It has literally been a decade since I did raw avr timers lol.

**Reply by kachurovskiy** [reply]
Hi @zyeborm, it's probably possible to drive the stepper with the timer but it surely won't be easy to account for all cases we have at hand. Feel free to give it a try and let us know how it went :)

**Reply by zyeborm** [reply]
I'm trying very hard not to take on another project right now ;-)
I was hoping it wouldn't actually change much of your code as you only seem to call step() in the one spot?
The change would be instead of calling step, update the desired position and let the PID loop take care of moving the stepper itself.
Mostly I was looking at it because the stepper drivers we wound up getting had 400 minimum steps per revolution on them lol, so increasing the step rate might be nice lol.

**Reply by kachurovskiy** [reply]
The key feature of NanoEls is that there's no error in stepper pulses, not even 1. Will we be able to maintain this guarantee with timers?

**Reply by zyeborm** [reply]
mmm, I dunno man. 
If you went to timers (and everything worked as I hoped it would) you could run high levels of micro stepping which can give much smoother movement, (the increase in accuracy goes down as you go over about 4x micro stepping, but the resonances and things do keep reducing) 
You'd also have time to use both channels of the encoder, probably on rising and falling edges so the resolution of your encoder position would go up quite a bit.
The following error would be a thing that you would then know about and you can define a limit on it. You might say max ferror of 8 steps. But if you have 16x micro stepping you are going to be actually more "accurate" at 8 steps out of place than you would be with full steps if it was "almost" time for a new step.
The other thing is you're purely reactive and *hard* in the current code I think? I'm still trying to wrap my head around it exactly, every encoder edge you pick up it looks like you then calculate X steps to send? But the speed has changed and now you're some behind, or possibly in front?
What happens when you run out of steps before the next encoder pulse comes? Do you pause and wait?

That design pattern I mentioned of running a PID loop on the stepper position is how linuxCNC works.
 
GRBL say they can do 30Khz on 3 axies with their setup, they use a bressenham line algorithm and a very tight ISR based loop that does their bit fiddling still in software. https://github.com/grbl/grbl/blob/master/grbl/stepper.c
Of course, they don't have the option to do hardware step generation because they need to run 3 axis at least.
You can see they are doing that trick of preloading the counter register on line 298 to get their frequency up btw.
It is nice that they are all integer based, no floats which are bad on an AVR and PID's are easier with floats.

Perhaps you might steal that pattern? It'd be less of an impact on your current code. I'd still probably do that trick with the fastPWM mode to get the hardware to put out a low, then go high for your step pulse time, finally go low again on counter rollover and drop you into the ISR where you could keep your existing "what do I do after I have made a step" type code and also preload the counter to get your next step time set right?

**Reply by kachurovskiy** [reply]
I gave it a try and for some reason I can't get the timer-based solution to run the motor faster than ~270 RPM - https://www.diffchecker.com/obLHOSkt/ - setting OCR1A to 1 or 100 or 500 makes no difference 👀

**Reply by zyeborm** [reply]
Oh sorry I didn't see this! I'll take a look at it asap

**Reply by zyeborm** [reply]
Ok having a quick look and bringing back some old memories.
First is stream of consciousness looking at the current code.

I have noticed a potential bug you might want to watch for, pendingPos is a multi byte global (as it needs to be really)
but you perform compare operations on it in your main loop and in the ISR, it's possible that the ISR will update pendingPos while the main loop is part way through an evaluation which might be "undesired" lol.
you can wrap your comparisons in code to block interrupts to avoid this. (sei,cei were the old school way I think?)

Your timer clock is prescaled (divided) by 8 in your setup (CS11 here) 
TCCR1B = (1 << WGM12) + (1 << CS11);
if you leave all the CS bits as 0 it'll run 8 times faster. I don't think you'll want to make use of the speed as such but by running faster and using the full 16 bits available you'll get higher resolution. (16Mhz/65535(16 bits) is 244hz slowest speed)

It looks like you're using arduino digitalread/write functions? I think wrapped behind your test harness. They are very slow (doesn't normally matter, but when you really want to go fast it can.
Take a look at https://roboticsbackend.com/arduino-fast-digitalwrite/#Why_Arduino_digitalWrite_is_not_fast
I'd be doing it using the method described here if only inside the ISR
https://roboticsbackend.com/arduino-fast-digitalwrite/#Modifying_only_one_pins_state 
However you may wind up putting pulses out too fast for the stepper driver to handle doing that.

If my maths is right you should be stepping at 4000Hz with the setup you have at the moment. 16000000/8/500 = 4000, if set to 100 you should get 20Khz. ISR rate. Are you perhaps exhausting pendingPos in your current code?


A different approach entirely and one that saves you from any bitbashing at all is
Change to fastPWM mode.
https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-7810-Automotive-Microcontrollers-ATmega328P_Datasheet.pdf

Pg 108
TCCR1A set COM1A1 and COM1A0 = 1 
Set OC1A/OC1B on compare match, clear OC1A/OC1B at BOTTOM (inverting mode)
This will use the timer compare outputs to drive the pin directly.

109
Set WGM to mode 14 (FastPWM with top defined by ICR1 so you can keep OCR1A as output compare match.)
(WGM is spread across a few registers)
Set ICR1 to SomeValue where SomeValue will define your step frequency
Set OCR1A to like SomeValue - 100(Or whatever you need it to be, the pin will go low at OCR1A, and high at ICR1 by itself)

In the OC1x interrupt you can set new values for OCR1A and ICR provided ICR (which isn't double buffered) isn't so small that the timer has already past it before it's run. I don't think it'll be an issue.
A way around that is to leave ICR as 65535 and OCR1A as 65000 (or whatever) then in the TOV1 (Timer overflow interrupt) just set a value for the timer. So you could say set the timer value to 30,000 and the next step pulse would happen twice as fast as if it were set to 0 to start counting up. 

Setting the value for the step frequency would probably be something dependant on the size of pendingPos, IE if pendingpos is larger then make the step frequency faster to catch up.
The disabling logic can probably remain the same. But for enable you probably want to force it to step immediately, so you'd set the timer value to OCR1A- 100 (or so) so then enable the timer so it'd trigger 100 cycles after you enable the timer. Probably an idea to wrap the timer set and interrupt enable in a SEI/CEI just so you don't have an encoder pulse ISR firing in between and messing everything up.

Got me interested in firing up a nano now and busting out the scope to see what's happening, but the rest of your code is a bit too big/detailed for me right now lol.
Do you use any real time chat platforms? I use signal if you'd like to chat in real time, it might be a lot quicker than this lol.

**Reply by kachurovskiy** [reply]
I'm using FastGPIO. I'm quite hesitant to invest more time into Nano timers as it's not been a good experience so far and debugging possibilities are non-existent. If you believe it can work then I'd welcome a proof of concept that respects soft limits and issues exactly the right amount of steps 😀 I know it's possible to run a stepper with PWM but it's not enough for the problem at hand.

Feel free to reach out via https://t.me/MaximKachurovskiy 😃

**Comment by zyeborm** [comment]
Ahh that's irritating, I really don't like how kinda crappy the inputs on the stepper drivers are generally. Like I get they are cheap but is differential really so hard?You'd probably get away with like 200 ohm which would still limit the short circuit current at least. BTW you can get zero ohm resistors. It's not uncommon to use them for stuff like this so you can have a resistor with a value on the board or change it out for 0 if you want to bypass it totally.

**Comment by zyeborm** [comment]
Just letting you know we got the first one running.
Would be nice to have that higher step rate 😜 but it's working pretty well so far.

**Reply by kachurovskiy** [reply]
![img](https://media.tenor.com/lLL0TK-IOjgAAAAC/i-would-like-to-see-it-smug.gif)

Higher stepper rate is in the works 😀

**Reply by zyeborm** [reply]
Have you changed things since that timer diff you posted above?
Kit.jpg looks interesting 😜

**Reply by kachurovskiy** [reply]
Thanks! No, didn't have any ideas how to improve the timer logic since that attempt.
